components/gitpod-protocol/go/gitpod-service.go

```diff
@@ -336,21 +338,51 @@ func (gp *APIoverJSONRPC) InstanceUpdates(ctx context.Context, instanceID string
 	chn := make(chan *WorkspaceInstance)
 
 	gp.mu.Lock()
-	if gp.subs == nil {
-		gp.subs = make(map[string]map[chan *WorkspaceInstance]struct{})
+	if gp.instanceSubs == nil {
+		gp.instanceSubs = make(map[string]map[chan *WorkspaceInstance]struct{})
 	}
-	if sub, ok := gp.subs[instanceID]; ok {
+	if sub, ok := gp.instanceSubs[instanceID]; ok {
 		sub[chn] = struct{}{}
 	} else {
-		gp.subs[instanceID] = map[chan *WorkspaceInstance]struct{}{chn: {}}
+		gp.instanceSubs[instanceID] = map[chan *WorkspaceInstance]struct{}{chn: {}}
 	}
 	gp.mu.Unlock()
 
 	go func() {
 		<-ctx.Done()
 
 		gp.mu.Lock()
-		delete(gp.subs[instanceID], chn)
+		delete(gp.instanceSubs[instanceID], chn)
+		close(chn)
+		gp.mu.Unlock()
+	}()
+
+	return chn, nil
+}
+
+// WorkspaceUpdates subscribes to workspace instance updates until the context is canceled
+func (gp *APIoverJSONRPC) WorkspaceUpdates(ctx context.Context, workspaceID string) (<-chan *WorkspaceInstance, error) {
+	if gp == nil {
+		return nil, errNotConnected
+	}
+	chn := make(chan *WorkspaceInstance)
+
+	gp.mu.Lock()
+	if gp.workspaceSubs == nil {
+		gp.workspaceSubs = make(map[string]map[chan *WorkspaceInstance]struct{})
+	}
+	if sub, ok := gp.workspaceSubs[workspaceID]; ok {
+		sub[chn] = struct{}{}
+	} else {
+		gp.workspaceSubs[workspaceID] = map[chan *WorkspaceInstance]struct{}{chn: {}}
+	}
+	gp.mu.Unlock()
+
+	go func() {
+		<-ctx.Done()
+
+		gp.mu.Lock()
+		delete(gp.workspaceSubs[workspaceID], chn)
```

Comment:
What happens if you have two `WorkspaceUpdates` subscriptions for the same workspace ID?

Here, I strongly suspect that the moment the first context cancels, it will break the other subscription also, that's because we're using a shared structure on the client. 

Could we instead return an object which provides the updates for a particular Workspace ID and avoid using shared state in the client?

Reply:
It's fine, the structure is a map of maps `map[string]map[chan *WorkspaceInstance]struct{}` using the channel as a key in the second map, channel is unique per every new call